sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,912 @@
+                        bytes_ptr: *mut u8,
+                        bytes_len: *mut u64,
+                        val: u64,
+                    ) -> u64;
+                }
+                let mut value = Vec::<u8>::with_capacity(4);
+                let mut value_len = 4u64;
+                unsafe {
+                    let result = sol_bignum_from_u32(
+                        value.as_mut_ptr() as *mut _ as *mut u8,
+                        &mut value_len as *mut _ as *mut u64,
+                        val as u64,
+                    );
+                    if result == 0 {
+                        value.set_len(value_len as usize);
+                        Ok(Self {
+                            is_negative: false,
+                            value,
+                        })
+                    } else {
+                        Err(result)
+                    }
+                }
+            }
+        }
+    }
+    /// Returns a BigNumber from a decimal string
+    pub fn from_dec_str(string: &str) -> Result<Self, u64> {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            match openssl::bn::BigNum::from_dec_str(string) {
+                Ok(bn) => Ok(Self {
+                    is_negative: bn.is_negative(),
+                    value: bn.to_vec(),
+                }),
+                Err(_) => Err(0x1),
+            }
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            extern "C" {
+                fn sol_bignum_from_dec_str(
+                    in_dec_str_ptr: *const u64,
+                    in_size: u64,
+                    out_ptr: *mut u8,
+                    out_size_ptr: *mut u64,
+                    out_is_neg_flag_ptr: *mut u64,
+                ) -> u64;
+            }
+            let mut value = Vec::<u8>::with_capacity(string.len());
+            let mut value_len = string.len() as u64;
+            let mut is_negative = 0u64;
+            unsafe {
+                let result = sol_bignum_from_dec_str(
+                    string.as_ptr() as *const _ as *const u64,
+                    string.len() as u64,
+                    value.as_mut_ptr() as *mut _ as *mut u8,
+                    &mut value_len as *mut _ as *mut u64,
+                    &mut is_negative as *mut _ as *mut u64,
+                );
+                if result == 0 {
+                    value.set_len(value_len as usize);
+                    Ok(Self {
+                        is_negative: is_negative != 0,
+                        value,
+                    })
+                } else {
+                    Err(result)
+                }
+            }
+        }
+    }
+
+    /// Flag indicating if negative or not
+    pub fn is_negative(&self) -> bool {
+        self.is_negative
+    }
+
+    /// Returns an array of bytes of self
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.value
+    }
+
+    /// Add BigNumbers
+    pub fn add(&self, rhs: &BigNumber) -> Result<Self, u64> {
```

Comment:
Not sure this works but what if `BigNumber` was also `#[cfg...` where if not bpf then it contained `BigNum` directly and only contained the byte array for BPF targets?

Reply:
I'll try it out, if conditional works it should be fine